### PR TITLE
(new) CIP-0001 & CIP-9999: Cardano Problem Statements

### PR DESCRIPTION
**Pull request by KtorZ**
This is a proposal to change CIP-0001 and rework how the CIP process currently function. This includes the introduction of a new type of document: Cardano Problem Statements (CPS); specified as part of CIP-9999 (read as _"CIP minus 1"_).

View [CIP-0001 Markdown](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-0001/README.md)

View [CIP-9999 Markdown](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-9999/README.md)

View [updated top level README](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/README.md)

### CODE REVIEW

File: CIP-9999/README.md
Comment (jehrhardt) [inline CIP-9999/README.md:11]:
```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/366
```

It looks like this is the correct link to the pull request for discussion.

Comment (KtorZ) [reply CIP-9999/README.md:11]: Indeed, I picked 365 in anticipation but there was a PR opened in the meantime!

**Review by rphair — APPROVED** [review]
Brilliant work all around 🤓

The proper lists (rather than comma-separated as before) of authors & discussions look much better in the header.  We should be on the lookout to be sure this different syntax also works when GitHub CIP `README.md` files are scraped onto `cips.cardano.org` and the Developer Portal.

I'm letting go of the comments I made before about having a template or example file visible on the front page for new visitors, writers, etc. to land on.  The links to each template are still difficult to spot, but people will find them if they know they're there.  The structure is also clear enough from the Specification > CIPs > [Structure](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-0001/README.md#structure) section that finding the CIP template isn't vital.

@KtorZ to facilitate agenda preparation for meetings, and quick review of relevant PRs during the meetings, can we have GitHub PR labels corresponding to these [Categories](https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-0001/README.md#categories)?  Preferably with the same prefix so they group together (e.g. `Category: Meta` instead of just `Meta`)?

In any case I think we would have to add an `RSS` (Rewards Sharing Schemes) category to this list.  Regardless of how any institution decides to respond to these proposals, after over 2 years of RSS proposals there are now several of them which will repeatedly need to be seen as a group, and for which none of the other listed Categories are applicable.

Also, did you have a process in mind for people with CIPs who decide to convert their CIP into a CPS?  For example CIP-0050 could be a good candidate for this (though I would happily leave that decision entirely to the author).

File: CIP-9999/README.md
Comment (michaelpj) [inline CIP-9999/README.md:85]: same comment as before

File: CIP-9999/README.md
Comment (michaelpj) [inline CIP-9999/README.md:90]: same comment about folder name

File: CIP-9999/README.md
Comment (michaelpj) [inline CIP-9999/README.md:112]: Given my previous comment about CPSs probably having multiple CIPs that "solve" parts of them, I wonder again about having something like "Addressed-by" instead. And perhaps the CPS could move to "Resolved" when it is completely addressed? Because that could also happen if it ceases to be a problem for some other reason.

File: CIP-9999/README.md
Comment (michaelpj) [reply CIP-9999/README.md:90]: Or just say "same as CIP-001" here also

File: CIP-9999/README.md
Comment (michaelpj) [inline CIP-9999/README.md:137]: "code owner" I think another new name for "project maintainer"?

File: CIP-9999/README.md
Comment (michaelpj) [inline CIP-9999/README.md:139]: Do we want to explicitly allow for the possibility of having multiple competing presentations of a problem? I guess we can leave it up to the editors.

**Review by michaelpj — COMMENTED** [review]
Looking good!

File: CIP-0001/README.md
Comment (rphair) [reply CIP-0001/README.md:54]: I also thing the collapsible sections don't work as well as straightforward markup would.  It's clear enough from the header division: while collapsing the text under each header can make it seem like the following text is associated with that header (e.g. the links to the template files are easily missed because they appear under "Licensing").

More of a problem: the collapsed sections _aren't searchable_.  People will be looking for keywords in that collapsed text (e.g. header field names or ecosystem tags) and won't be able to find them unless those sections are expanded.

File: CIP-0001/README.md
Comment (rphair) [reply CIP-0001/README.md:98]: I've been asking authors to use a directory name of `CIP-` followed by a very abbreviated name of their proposal.

We should also tell submitters to provide a link to this file in the OP of the PR thread.  Any time we create the link for them (as we have after submission or during triage), we're inviting them to change the README pathname and have this link become invalid.  If they create it, generally they will maintain it.

The place to tell them to create this link is probably here, though; not in the text above: https://github.com/cardano-foundation/CIPs/blob/cip-cps-rework/CIP-0001/README.md#1a-authors-open-a-pull-request

File: CIP-0001/README.md
Comment (KtorZ) [reply CIP-0001/README.md:68]: I originally put addresses there, and thought it might be confusing with the other meaning of addresses in our space. People might misinterpret that as a kind of donation section or whatever. I preferred to avoid the term altogether.

File: CIP-0001/README.md
Comment (michaelpj) [inline CIP-0001/README.md:146]: Pondering this in the case of Plutus builtins in particular, and I was wondering about insisting that CIPs proposing Plutus builtins be required to include test cases, or more generally a test plan. Because often the proposer is the one who actually knows what it's supposed to do, so they should include them here.

That made me wonder whether it's something that we want to include in _any_ CIP: the acceptance criteria SHOULD include test cases that must pass, and the implementation plan SHOULD include a testing plan.

WDYT?

File: CIP-0001/README.md
Comment (crptmppt) [reply CIP-0001/README.md:11]: having been the one who put CIP-0001 together and shouldered much of the initial weight of the CIP framework, I want to make sure I remain listed as author, thanks!

**Review by crptmppt — DISMISSED** [review]
as noted in the comment, changing the authors of a CIP is not appropriate. I kindly requests the original authors be kept and that Michael be added as alternative.

**Comment by rphair** [comment]
Despite practically every sentence being changed in the last draft, I would still support retaining @crptmppt (and maybe others) as an author in the header because of the practical work that went into establishing and documenting the original process.

I don't see how the current evolution of the CIP process which @KtorZ has documented much more rigorously today could never have happened except as experimental revisions from the old process: both in practice and in the earlier revisions of the document.

The paragraphs & sections in common between old & new documents may have changed completely in text, but they still serve the same purpose as the original.  Having expressed those concepts in the first place is just as much a part of this living document as the current text itself.

File: CIP-0001/README.md
Comment (crptmppt) [reply CIP-0001/README.md:54]: +1 for removing the collapsing sections, prefer the entire document scrolleable

File: .github/CIP-TEMPLATE.md
Comment (crptmppt) [inline .github/CIP-TEMPLATE.md:5]: "Proposed"

File: CIP-0001/README.md
Comment (crptmppt) [inline CIP-0001/README.md:12]: While this expands on the original CIP, I think the date should be that of the initial proposal here. It might be confusing to some to see dates all over the place since the process is functionally the same

File: CIP-0001/README.md
Comment (KtorZ) [reply CIP-0001/README.md:201]: I do. Will clarify.

File: CIP-0001/README.md
Comment (KtorZ) [reply CIP-0001/README.md:240]: Had to be in some places, thought it was the least worse.

File: CIP-9999/README.md
Comment (KtorZ) [reply CIP-9999/README.md:112]: There already is a `Proposed Solutions` field in the preamble. Having another `Addressed-By` field feels redundant since, any `Proposed Solutions` in an `Active` would be effectively addressing (even partially) the problem. If you're suggesting to rename `Proposed Solutions` to `Addressed-By`, that's an option. I have no strong feeling about naming here.

**Comment by rphair** [comment]
I'm still happy to go ahead with this, especially now that https://github.com/cardano-foundation/CIPs/pull/366#pullrequestreview-1170652027 has been tentatively resolved by restoring @crptmppt  to the author list.  I am guessing the red flag might still remain because @SebastienGllmt @dcoutts  haven't been similarly restored?

If this is still a potential blocker, I saw on [EIP-1](https://eips.ethereum.org/EIPS/eip-1) the [EIP Editors](https://eips.ethereum.org/EIPS/eip-1#eip-editors) section has a list of "Emeritus EIP editors" to indicate substantial prior contributions.  Maybe @KtorZ that would work for CIP-0001 (e.g. "Emeritus authors") if that's a way to go forward that places everyone where they want to be.  This could work for other evolving CIPs as well.

File: CIP-9999/README.md
Comment (michaelpj) [reply CIP-9999/README.md:108]: I guess it just seems weird to me to put the CPS in the Solved state if only part of the problem has been addressed. That seems misleading. But then, leaving it in Open when it has been partially solved is perhaps also misleading. So I don't know what to do.

File: CIP-0001/README.md
Comment (michaelpj) [reply CIP-0001/README.md:168]: I guess more generally I'm thinking that Acceptance Criteria usually involve some expectations of behaviour change, and it's not unreasonable to expect the proposer to at least broadly sketch out the expected behaviour changes.

e.g. for reference inputs we could have asked me to provide some transaction submission scenarios that would now work. That then is very helpful for generating test cases.

**Comment by michaelpj** [comment]
Personally, I lean towards the liberal attitude towards authorship: offering credit and thanks is cheap, so the bar for doing it should be low. IMO it costs us nothing to have everyone who ever made a significant contribution to the document listed as an author.

**Comment by KtorZ** [comment]
Not sure why everyone keeps arguing about authorship, as I've said, I have no particular wish to _remove_ authors. The document was simply written from scratch and I "scrapped" the old one, with its authors. The rationale being that I don't want to put words in other people's mouth; and none of the previous CIP-0001 authors have contributed to this new document. 

@crptmppt indicated he would like to be preserved in the authors list for the original work on CIP-0001, so I added him back. I will add back @dcoutts and @SebastienGllmt on this new document if they wish to.

File: CIP-9999/README.md
Comment (KtorZ) [reply CIP-9999/README.md:108]: Then I need to clarify something because, `Solved` is only assigned to problems that are **fully** solved. A problem that is only partially solved remains in `Open`, but is updated accordingly to show what's left to solve. 

As @abailly-iohk also hinted once, it's probably better to have CPS be focused on a single "atomic" problem as much as possible to avoid this type of situation.

Comment (michaelpj) [reply CIP-9999/README.md:108]: Well, the text says just:

> Problems for which one or more solutions have been found and implemented. 

Which certainly has left me confused about the partial solution state.

> As @abailly-iohk also hinted once, it's probably better to have CPS be focused on a single "atomic" problem as much as possible to avoid this type of situation.

I don't think I agree with this. I think it's very unlikely that the "problem view" lines up with the "solution view" so neatly. In practice we often need to do many things to fully address a user-visible problem!

Comment (abailly-iohk) [reply CIP-9999/README.md:108]: The important part is 
> as much as possible
Of course, real life is messy and problems do not usually align neatly with solutions, but we should encourage definitions of problems that are actionable and "compact", eg. instead of a CPS trying to "Solve world hunger", write a CPS saying "How to salvage France's hypermarkets food trashed?" 
Not sure this is a great example but I guess the idea is there

**Comment by ccgarant** [comment]
**CIP/CPS part numbering scheme** 
Since this is a new revision, just something to maybe consider. I haven't seen this done in crypto land, but in engineering land we have easily identifiable drawing and part numbers, like XXX-YYY-ZZZ-AAA format.
E.g. 901-301-001-101
 - 901 is the platform
 - 301 is a hydraulic part
 - 001 is the specific part
 - 101 is the revision number (e.g. 103 is the next rev)

Maybe a new, helpful, easy to visually check numbering scheme could be:
- CIP-META-0021B, where B is rev 2
- CPS-WALLET-0003A
- MBO-RSS-0001D

This might cause renumbering/mapping of old CIPs which might be problematic, so maybe:
 - CIP-0068-W-1, where W is for Wallet, and 1 is for Rev 1? (or CIP-0068-WALLET-01, MBO-0001-RSS-02, etc.)

Just food for thought and consideration...

**Comment by abailly-iohk** [comment]
I have a meta  question about CIPs: I am having a hard time distinguishing the _Motivation_ section from the _Rationale_ section. According to dictionaries:
> **Motivation**: a reason or reasons for acting or behaving in a particular way.
> **Rationale**: a set of reasons or a logical basis for a course of action or belief.

These seems pretty much synonymous to me :thinking_face:

**Comment by abailly-iohk** [comment]
I am really having a hard time wrapping my head around what `Active` really means, and how to get there. I wonder if having some form of  state machine diagram in the body of the CIP document would not help?

**Comment by KtorZ** [comment]
@abailly-iohk there used to be a state-machine diagram, but it's basically as simple as:

```mermaid
stateDiagram-v2
    [*] --> Proposed
    Proposed --> Active
    Proposed --> Inactive
```

Hence why I don't much see the value to be frank. Is there anything I could add to make the definition of what `Active` means clearer? We've iterated several times on that and I thought that last draft was pretty clear:

> An _'Active'_ CIP has taken effect according to the criteria defined in its _'Path to Active'_ section.
> What that means depends on the nature of the CIP, typically:
>
> - For CIPs that relate to particular projects or pieces of technology, it becomes _'Active'_ by being implemented and released;
> - For changes to the Cardano protocol, a CIP will become _'Active'_ by being live on the Cardano mainnet;
> - For ecosystem standards, it means gaining sufficient and visible adoption in the community.

**Comment by abailly-iohk** [comment]
Yeah, I agree it's a bit silly... I don't know why but it cannot stick to my mind, perhaps because as you mention it depends on the actual CIP 🤔

**Comment by KtorZ** [comment]
Perhaps we have focused too much on how to become active but should also explain what are the implications of being 'Active'.

'Active' means completed. It's kind of analogous to a release for a software. The proposal has matured, was challenged and has demonstrated to be solid enough to be followed. 

On the CIP process itself, it means that we would generally not substantially change any 'Active' CIP but instead, ask people to make a new one. While still in 'Proposed

**Comment by KtorZ** [comment]
(sorry, pressed the 'Close with comment' button by accident).

**Comment by abailly-iohk** [comment]
But it seems to me there's a missing step between `Proposed` and `Active` which should be something like `Approved` or `Accepted` or `Planned` or whatever. You can make a CIP which is structurally sound, well-written, is vouched by CIP editors, becomes `Proposed` but ultimately receives more attention, leads to prototypes, and is rejected because it's too complicated, or noone actually cares to implement this. 
Sorry, it seems I am reenacting an old debate here so feel free to ignore my comment and confusion.

File: CIP-9999/README.md
Comment (michaelpj) [reply CIP-9999/README.md:108]: > "How to salvage France's hypermarkets food trashed?"

... which is hardly going to have one solution, no? :) I just think it's unrealistic to suppose that every CPS will have one solution. In fact, my bet would be that that's the less common case.

Comment (abailly-iohk) [reply CIP-9999/README.md:108]: That a problem stated in a CPS has more than one solution is certainly desirable and that's actually what I was, awkwardly, trying to express.

Comment (abailly-iohk) [reply CIP-9999/README.md:108]: The idea I was trying to convey is that a CPS should express a problem that has  _at least_ one solution which itself could be expressed as a CIP. We want to rule out problems whose solution required something that would not fit in a single CIP (I think).

**Comment by KtorZ** [comment]
@abailly-iohk -> I have slightly reworded the 'Status: Active' section as such: 

---

##### Status: Active

An _'Active'_ CIP has taken effect according to the criteria defined in its _'Path to Active'_ section. Said differently, each CIP defines by which criterias it can become _'Active'_ and those criteria are included in the review process. Exact criteria thereby depends on the nature of the CIP, typically:

- For CIPs that relate to particular projects or pieces of technology, it becomes _'Active'_ by being implemented and released;
- For changes to the Cardano protocol, a CIP becomes _'Active'_ by being live on the Cardano mainnet;
- For ecosystem standards, it means gaining sufficient and visible adoption in the community.

A proposal that is _'Active'_ is considered complete and is synonymous with "production readiness" when it comes to the maturity of a solution. _'Active'_ CIPs will not be updated substantially (apart from minor edits, proofreading and added precisions). They can, nevertheless, be challenged through new proposals if need be.

**Comment by rphair** [comment]
> @abailly-iohk ... distinguishing the Motivation section from the Rationale section...
>
>  - Motivation: a reason or reasons for acting or behaving in a particular way.
>  - Rationale: a set of reasons or a logical basis for a course of action or belief.
>
> These seems pretty much synonymous to me...

They are: but only without respect to the progression of _time_.  In our process, the motivation _precedes_ the evolution of the proposal: so whatever past set of conditions that proposal hopes to improve will otherwise remain in force.  _After_ we establish that we're doing something about it (as outlined in the proposal), the rationale describes what we'd be doing in the present & future.

Also: in general they do mean the same thing if they are _about_ the same thing.  But in the case of CIP's, the "motivation" is for the proposal itself, while the "rationale" is for _the method described in_ the proposal.

**Review by crptmppt — APPROVED** [review]
I changed the authors list to the original authors + added MPJ (for the changes going into 0001). As the CIP process hardly changes beside the CPS addition I feel it is fine to merge with the added change. If there are issues regarding the addended authorship on this PR, I suggest this be submitted as a different CIP #, which can override 0001 (this could then be noted in the body of 0001) - "convenience" should not override facts, having created 0001 and the entire original CIP process, I request original authorship of 0001 be noted in the PR.  Other CIP fine as-is.

**Pull request merged**
